File: year2/mpp/java/MotorcycleContestHibernate/MotorcycleServer/src/main/java/networking/MotorcycleClientRpcWorker.java
package networking;

import domain.Participant;
import domain.Race;
import domain.User;
import dtos.CredentialsDto;
import dtos.ParticipantDto;
import dtos.RaceDto;
import dtos.UserDto;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import services.MotorcycleObserver;
import services.MotorcycleService;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class MotorcycleClientRpcWorker implements Runnable, MotorcycleObserver {
    private static final Logger LOGGER = LogManager.getLogger(MotorcycleClientRpcWorker.class.getName());
    private final MotorcycleService service;
    private final Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public MotorcycleClientRpcWorker(MotorcycleService service, Socket connection) {
        LOGGER.traceEntry("creating worker for server " + service + " and connection " + connection);
        this.service = service;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            LOGGER.warn("output stream failed");
        }
        LOGGER.traceExit("worker successfully created");
    }

    public void run() {
        while (connected) {
            try {
                Object request = input.readObject();
                LOGGER.info("received request " + request);
                Response response = handleRequest((Request) request);
                LOGGER.info("response for request " + response);
                if (response != null) {
                    LOGGER.info("sending response " + response);
                    sendResponse(response);
                }
            } catch (IOException | ClassNotFoundException e) {
                LOGGER.warn("request failed " + e);
            }
        }
        try {
            LOGGER.info("closing connection");
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            LOGGER.warn("closing connection failed " + e);
        }
    }

    private static final Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private static final Response errorResponse = new Response.Builder().type(ResponseType.ERROR).build();

    private Response handleRequest(Request request){
        LOGGER.traceEntry("entering handleRequest with " + request);
        Response response = null;
        if (request.type() == RequestType.LOGIN){
            LOGGER.traceEntry("resolving login request ");
            UserDto userDto = (UserDto) request.data();
            try {
                LOGGER.info("finding user with username {} and password {}", userDto.getUsername(), userDto.getPassword());
                Optional<User> user = service.authenticate(userDto.getUsername(), userDto.getPassword());
                if (user.isEmpty()) {
                    LOGGER.traceExit("user not found");
                    return okResponse;
                }
                service.addMotorcycleObserver(this);
                LOGGER.traceExit("exiting with user " + user);
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(user.get())
                        .build();
            } catch (Exception e) {
                LOGGER.warn("finding user failed");
                connected = false;
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        else if (request.type() == RequestType.GET_RACES){
            LOGGER.traceEntry("resolving get races request ");
            try {
                LOGGER.info("finding races");
                List<Race> races = StreamSupport
                        .stream(service.getRaces().spliterator(), false)
                        .collect(Collectors.toList());
                LOGGER.traceExit("exiting with races {}", races);
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(races)
                        .build();
            } catch (Exception e) {
                LOGGER.warn("finding races failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        else if (request.type() == RequestType.SEARCH_RACES){
            LOGGER.traceEntry("resolving get races information request ");
            try {
                LOGGER.info("finding races information");
                List<RaceDto> races = service.getRaceInfos();
                LOGGER.traceExit("exiting with races dtos {}", races);
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(races)
                        .build();
            } catch (Exception e) {
                LOGGER.warn("finding races failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        else if (request.type() == RequestType.FILTER_RACES){
            LOGGER.traceEntry("resolving get race by capacity request ");
            try {
                int capacity = (int) request.data();
                LOGGER.info("finding race by capacity {}", capacity);
                RaceDto race = service.getRaceByCapacity(capacity);
                LOGGER.traceExit("exiting with race dto {}", race);
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(race)
                        .build();
            } catch (Exception e) {
                LOGGER.warn("finding race by capacity failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        else if (request.type() == RequestType.GET_PARTICIPANTS){
            LOGGER.traceEntry("resolving get participants by team request ");
            try {
                String teamName = (String) request.data();
                LOGGER.info("finding participants by team {}", teamName);
                List<ParticipantDto> participants = service.getRaceParticipantsByTeam(teamName);
                LOGGER.traceExit("exiting with participant dtos {}", participants);
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(participants)
                        .build();
            } catch (Exception e) {
                LOGGER.warn("finding participants by team failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        else if (request.type() == RequestType.REGISTER){
            LOGGER.traceEntry("resolving register participant request ");
            try {
                CredentialsDto credentials = (CredentialsDto) request.data();
                service.registerParticipant(credentials.getName(), credentials.getTeamName(), credentials.getCapacity());
                LOGGER.traceExit("register participant successful");
                return okResponse;
            } catch (Exception e) {
                LOGGER.warn("finding participants by team failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        else if (request.type() == RequestType.LOGOUT){
            UserDto userDto = (UserDto) request.data();
            try {
                connected = false;
                service.removeMotorcycleObserver(this);
                return okResponse;

            } catch (Exception e) {
                LOGGER.warn("logout failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        return response;
    }

    private void sendResponse(Response response) throws IOException{
        LOGGER.traceEntry("sending response {} ", response);
        output.writeObject(response);
        output.flush();
        LOGGER.traceExit("sending response {} successful", response);
    }

    @Override
    public void registeredParticipant(Participant participant) {
        LOGGER.info("sending registered participant response");
        Response response = new Response.Builder().type(ResponseType.REGISTERED_PARTICIPANT).data(participant).build();
        try {
            sendResponse(response);
        } catch (IOException e) {
            LOGGER.info("sending registered participant failed");
            e.printStackTrace();
        }
    }
}

File: year2/mpp/c#/MotorcycleClientServer/MotorcycleServer/Networking/MotorcycleClientRpcWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;

using MotorcycleCommon.Networking;
using MotorcycleCommon.Services;
using MotorcycleContest.Domain.Dtos;
using MotorcycleContest.Domain.Entities;

namespace MotorcycleServer.Networking
{
    public class MotorcycleClientRpcWorker : IMotorcycleObserver
    {
        // protected static readonly ILog Logger = LogManager.GetLogger("MotorcycleClientRpcWorker");
        
        private IMotorcycleService service;
        private TcpClient connection;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;
        
        public MotorcycleClientRpcWorker(IMotorcycleService service, TcpClient connection)
        {
            // Logger.Info("creating motorcycle rpc client worker");
            this.service = service;
            this.connection = connection;
            try
            {
                // Logger.Info("initiating connection");
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception exception)
            {
                // Logger.Warn("initiating connection failed");
                Console.WriteLine(exception.StackTrace);
            }
        }

        public virtual void Run()
        {
            // Logger.Info("running worker thread");
            while (connected)
            {
                try
                {
                    // Logger.Info("getting request");
                    object request = formatter.Deserialize(stream);
                    // Logger.Info("handling request " + request.ToString());
                    object response = HandleRequest((Request) request);
                    if (response != null)
                    {
                        // Logger.Info("sending response " + response.ToString());
                        SendResponse((Response) response);
                    }
                }
                catch (Exception exception)
                {
                    if (connected)
                    {
                        // Logger.Warn("running worker stopped with exception " + exception.Message);
                    }
                    break;
                }
            }

            try
            {
                // Logger.Info("closing connection");
                stream.Close();
                connection.Close();
            }
            catch (Exception exception)
            {
                // Logger.Warn("closing connection failed with exception " + exception.Message);
            }
        }

        private static Response okResponse = new Response() { Type = ResponseType.OK };
        private static Response errorResponse = new Response() { Type = ResponseType.ERROR };

        private Response HandleRequest(Request request)
        {
            // Logger.InfoFormat("handling request {0}", request);
            Response response = null;
            switch (request.Type)
            {
                case RequestType.LOGIN:
                {
                    // Logger.Info("resolving login request");
                    UserDto userDto = (UserDto) request.Data;
                    try
                    {
                        // Logger.InfoFormat("finding user with username {0} and password {1}", username, password);
                        User user = service.Authenticate(userDto.Username, userDto.Password);
                        if (user == null)
                        {
                            // Logger.Info("user not found");
                            return okResponse;
                        }
                        service.AddMotorcycleObserver(this);
                        // Logger.Info("exiting with user {0}", user);
                        return new Response()
                        {
                            Type = ResponseType.OK,
                            Data = user
                        };
                    }
                    catch (Exception)
                    {
                        // Logger.Warn("finding user failed");
                        connected = false;
                        return errorResponse;
                    }

                    break;
                }
                case RequestType.GET_RACES:
                    // Logger.Info("resolving get races request");
                    try
                    {
                        // Logger.Info("finding races");
                        List<Race> races = service.GetRaces().ToList();
                        // Logger.InfoFormat("found races {0}", races);
                        return new Response()
                        {
                            Type = ResponseType.OK,
                            Data = races
                        };
                    }
                    catch (Exception exception)
                    {
                        // Logger.Warn("finding races failed");
                        return errorResponse;
                    }

                    break;
                case RequestType.FILTER_RACES:
                    // Logger.Info("resolving get race by capacity request");
                    try
                    {
                        int capacity = (int) request.Data;
                        // Logger.InfoFormat("finding race by capacity {0}", capacity);
                        RaceDto race = service.GetRaceByCapacity(capacity);
                        // Logger.InfoFormat("found race {0}", race);
                        return new Response()
                        {
                            Type = ResponseType.OK,
                            Data = race
                        };
                    }
                    catch (Exception exception)
                    {
                        // Logger.Warn("finding races failed");
                        return errorResponse;
                    }

                    break;
                case RequestType.GET_PARTICIPANTS:
                    // Logger.Info("resolving get participants request");
                    try
                    {
                        String teamName = (String) request.Data;
                        // Logger.InfoFormat("finding participants by team {0}", teamName);
                        List<ParticipantDto> participants = service.GetRaceParticipantsByTeam(teamName);
                        // Logger.InfoFormat("exiting with participant dtos {0}", participants);
                        return new Response()
                        {
                            Type = ResponseType.OK,
                            Data = participants
                        };
                    }
                    catch (Exception exception)
                    {
                        // Logger.Warn("finding participants failed");
                        return errorResponse;
                    }

                    break;
                case RequestType.REGISTER:
                    // Logger.Info("resolving register participant request");
                    try
                    {
                        CredentialsDto credentials = (CredentialsDto) request.Data;
                        service.RegisterParticipant(credentials.Name, credentials.TeamName, credentials.Capacity);
                        // Logger.Info("register participant");
                        return okResponse;
                    }
                    catch (Exception exception)
                    {
                        // Logger.Warn("finding participants failed");
                        return errorResponse;
                    }

                    break;
                case RequestType.LOGOUT:
                    // Logger.Info("resolving logout request");
                    try
                    {
                        connected = false;
                        service.RemoveMotorcycleObserver(this);
                        // Logger.Info("logout successful");
                        return okResponse;
                    }
                    catch (Exception exception)
                    {
                        // Logger.Warn("logout failed");
                        return errorResponse;
                    }

                    break;
                default:
                    return response;
            }
        }

        private void SendResponse(Response response)
        {
            // Logger.InfoFormat("sending response {0}", response);
            formatter.Serialize(stream, response);
            stream.Flush();
        }

        public void RegisteredParticipant(Participant participant)
        {
            // Logger.InfoFormat("registered participant {0}", participant);
            Response response = new Response()
            {
                Type = ResponseType.REGISTERED_PARTICIPANTS,
                Data = participant
            };
            try
            {
                SendResponse(response);
            }
            catch (Exception exception)
            {
                // Logger.Warn("sending registered participant failed");
                Console.WriteLine(exception.StackTrace);
            }
        }
    }
}
